replace Long objects with Trove

abyrd opened the pull request:
This demonstrates replacing all the autoboxed Longs with Trove primitive collections. This should use less memory, cause less garbage collection, and be faster.

With Trove loading Norway OSM takes 24 seconds.
With Autoboxing it takes 35 seconds.
Trove seems to offer a 1/3 speedup.

Both hover at about 5GB heap during load, but with 12GB max so this needs to be tested better.

One potential problem is that some methods with primitive return types in Trove will return zero to indicate "no value". We would need to make sure those methods are not used, or used properly.

I do think I found and fixed a bug while working on this: in Area, right after comment "Precheck for impossible situations", action is taken on a list that never has anything added to it.

t2gran commented on src/main/java/org/opentripplanner/graph_builder/module/osm/OSMDatabase.java:
```diff
@@ -645,7 +650,8 @@ private void processMultipolygonRelations() {
                 }
-                for (Long nodeId : way.getNodeRefs()) {
+                for (TLongIterator iter = way.getNodeRefs().iterator(); iter.hasNext(); ) {
+                    long nodeId = iter.next();
```
I think a while loop reads better in this case - but no big deal.

abyrd replied:
True, the while loop itself reads better, but a while loop leaves the iterator in the containing scope. I like that the iterator goes out of scope when the for loop completes. The ideal would be `for (int nodeId : way.getNodeRefs().asArray())` but unfortunately Trove prevents you from seeing the internal array in a `TIntArrayList`.

abyrd replied:
I guess this is the idiomatic Trove way:
```
                boolean allNodesPresent = way.getNodeRefs().forEach(nodeId -> {
                    if (nodesById.containsKey(nodeId)) {
                        return false;
                    } else {
                        MapUtils.addToMapSet(areasForNode, nodeId, way);
                        return true;
                    }
                });
                if (!allNodesPresent) {
                    // this area is missing some nodes, perhaps because it is on
                    // the edge of the region, so we will simply not route on it.
                    continue RELATION;
                }
```
Which is even more verbose. I just changed it to a while loop, the scope only continues about 5 lines past the end of the loop.

t2gran commented:
Good work, the improvement achieved here will save me some minutes every day.

abyrd commented:
Thanks for the approval. I am going to do one more read-through to check whether we ever might get zeros indicating "no value", then merge.

abyrd commented:
I reviewed this one more time because Trove collections are not an exact drop-in replacement for Java collections. Some methods return primitives, so cannot return null to indicate a missing value. I wanted to check that the revised code did not depend on any null "missing value" return values.

In OSMDatabase all the TLongObjectMaps are fine, they can return non-primitive nulls from get(). It's the TLongSets and any primitive-valued maps that can be problematic because they return primitives. However the TLongSets are always used with .contains(), not by get() followed by a null check, so they should be fine. There doesn't seem to be any checking for missing values in any map get() calls either.

I did see that DisjointSet was still using collections of Integer objects instead of TIntList and TObjectIntMap. I added a commit updating it to use Trove. Unfortunately the DisjointSet#find() method does look for missing elements using a get followed by a null check. This migration was pretty straightforward though.

Working on this did bring to my attention that the DisjointSet class has some relatively cryptic logic using negative numbers and negated variables in size calculations, with no comments or references on how it works. I could probably work out what it's doing, but we should step back and ask the question of whether we want to maintain this code in the long term. It is only used by AreaGroup, which is primarily used in visibility calculations, which are themselves handled by a largely undocumented port of an external library, whose comments say "this port undoubtedly introduced a number of bugs (and removed some features)."

We can probably come up with a simpler and more maintainable way to handle areas, and better connect them to ways and vertices representing things like stairs that are floating inside the areas. See #1530, where I'll post a few more thoughts on this.

I think this PR is now ready to merge though. @t2gran I'll give you a bit to take another look at it since I've added commits and changed the base, but given how minor the changes are if I don't hear anything from you I'll just merge, on the basis of your previous approving review.
